ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveRewriteToDataSketchesRule.java

```diff
@@ -0,0 +1,327 @@
+        projRex = rexBuilder.makeCall(projectOperator, ImmutableList.of(projRex));
+        projRex = rexBuilder.makeCall(SqlStdOperatorTable.ROUND, ImmutableList.of(projRex));
+        projRex = rexBuilder.makeCast(origType, projRex);
+
+        newAggCalls.add(newAgg);
+        newProjects.add(projRex);
+      }
+
+    }
+
+    class PercentileContRewrite extends RewriteProcedure {
+
+      public PercentileContRewrite(String sketchClass) {
+        super(sketchClass);
+      }
+
+      @Override
+      boolean isApplicable(AggregateCall aggCall) {
+        // FIXME: also check that args are: ?,?,1,0 - other cases are not supported
+        return !aggCall.isDistinct() && aggCall.getArgList().size() == 4
+            && aggCall.getAggregation().getName().equalsIgnoreCase("percentile_cont") && !aggCall.hasFilter();
+      }
+
+      @Override
+      void rewrite(AggregateCall aggCall) {
+        RelDataType origType = aggregate.getRowType().getFieldList().get(newProjects.size()).getType();
+
+        Integer argIndex = aggCall.getArgList().get(1);
+        RexNode call = rexBuilder.makeInputRef(aggregate.getInput(), argIndex);
+
+        RelDataType floatType = rexBuilder.getTypeFactory().createSqlType(SqlTypeName.FLOAT);
+        call = rexBuilder.makeCast(floatType, call);
+        newProjectsBelow.add(call);
+
+        ArrayList<Integer> newArgList = Lists.newArrayList(newProjectsBelow.size() - 1);
+
+        SqlAggFunction aggFunction = (SqlAggFunction) getSqlOperator(DataSketchesFunctions.DATA_TO_SKETCH);
+        boolean distinct = false;
+        boolean approximate = true;
+        boolean ignoreNulls = true;
+        List<Integer> argList = newArgList;
+        int filterArg = aggCall.filterArg;
+        RelCollation collation = aggCall.getCollation();
+        RelDataType type = rexBuilder.deriveReturnType(aggFunction, Collections.emptyList());
+        String name = aggFunction.getName();
+
+        AggregateCall newAgg = AggregateCall.create(aggFunction, distinct, approximate, ignoreNulls, argList, filterArg,
+            collation, type, name);
+
+        Integer origFractionIdx = aggCall.getArgList().get(0);
+        RexNode fraction = getProject(aggregate.getInput()).getChildExps().get(origFractionIdx);
+        fraction = rexBuilder.makeCast(floatType, fraction);
+
+        SqlOperator projectOperator = getSqlOperator(DataSketchesFunctions.GET_QUANTILE);
+        RexNode projRex = rexBuilder.makeInputRef(newAgg.getType(), newProjects.size());
+        projRex = rexBuilder.makeCall(projectOperator, ImmutableList.of(projRex, fraction));
+        projRex = rexBuilder.makeCast(origType, projRex);
+
+        newAggCalls.add(newAgg);
+        newProjects.add(projRex);
+
+      }
+
+    }
+
+    private Project getProject(RelNode input) {
+      if (input instanceof Project) {
+        return (Project) input;
+      }
+      if (input instanceof HepRelVertex) {
```

Comment:
in case of percentile_cont we must have a Project there ; to inject the static arguments of the percentile_cont function

did you meaned something like:
```
    super(operand(HiveAggregate.class, operand(HiveProject.class, any())));
```
for `count(distinct)` the restiriction to have a project there doesn't exists...so I think this would mean I should create separate rules for every rewrite - right? I started with that approach then later merged them into one - but they might be better of separately....

Reply:
I see. I think you can keep them in a single one (certainly in this patch). For the matching, you can match on `RelNode.class` then and use `instanceof` as usual; the idea is to not rely on the node type within the rule logic.
```
    super(operand(HiveAggregate.class, operand(RelNode.class, any())));
```

Reply:
ok - I've update it - I'll separate the 3 rules in next patch differently